src/Pact/Types/Term.hs

```diff
@@ -324,14 +324,18 @@ data FunApp = FunApp {
 instance HasInfo FunApp where getInfo = _faInfo
 
 -- | Variable type for an evaluable 'Term'.
-data Ref =
+data Ref' d =
   -- | "Reduced" (evaluated) or native (irreducible) term.
-  Direct (Term Name) |
+  Direct d |
   -- | Unevaulated/un-reduced term, never a native.
-  Ref (Term Ref)
-  deriving (Eq,Show)
+  Ref (Term (Ref' d))
+  deriving (Eq,Show,Functor,Foldable,Traversable,Generic)
```

Comment:
This is exactly `data Free f a = Pure a | Free (f (Free f a))`, so lets try dropping in a `Free` library in the future